### ISSUE TITLE
Parallel writes to new pucket contend over descriptor creation

### ISSUE THREAD

**Issue by cdagraca** [intenthq/pucket#16]
Pucket's `findOrCreate` checks for the existence of the descriptor, and attempts to create it if not found. However, if writing in parallel (e.g. Spark output with multiple partitions), each task checks, finds no descriptor, and tries to write one. The first one succeeds. The rest fail because the file now _does_ exist. How much of a problem this poses depends on whether the calling framework retries failed write tasks.

**Comment by cdagraca** [comment]
Error logged will look something like this:
```
17/08/21 09:17:27 WARN scheduler.TaskSetManager: Lost task 1.1 in stage 1.0 (TID 620, ip-10-10-8-177.eu-west-1.compute.internal, executor 2): org.apache.hadoop.ipc.RemoteException(org.apache.hadoop.hdfs.protocol.AlreadyBeingCreatedException): Failed to CREATE_FILE /tmp/my_dir/sub_dir/.pucket.descriptor for DFSClient_NONMAPREDUCE_-1698245061_101 on 10.10.8.177 because this file lease is currently owned by DFSClient_NONMAPREDUCE_1067292598_94 on 10.10.8.177
        at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.recoverLeaseInternal(FSNamesystem.java:2931)
        at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.startFileInternal(FSNamesystem.java:2561)
        at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.startFileInt(FSNamesystem.java:2450)
        at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.startFile(FSNamesystem.java:2334)
        at org.apache.hadoop.hdfs.server.namenode.NameNodeRpcServer.create(NameNodeRpcServer.java:624)
        at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolServerSideTranslatorPB.create(ClientNamenodeProtocolServerSideTranslatorPB.java:397)
        at org.apache.hadoop.hdfs.protocol.proto.ClientNamenodeProtocolProtos$ClientNamenodeProtocol$2.callBlockingMethod(ClientNamenodeProtocolProtos.java)
        at org.apache.hadoop.ipc.ProtobufRpcEngine$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine.java:616)
        at org.apache.hadoop.ipc.RPC$Server.call(RPC.java:982)
        at org.apache.hadoop.ipc.Server$Handler$1.run(Server.java:2049)
        at org.apache.hadoop.ipc.Server$Handler$1.run(Server.java:2045)
        at java.security.AccessController.doPrivileged(Native Method)
        at javax.security.auth.Subject.doAs(Subject.java:422)
        at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1698)
        at org.apache.hadoop.ipc.Server$Handler.run(Server.java:2043)

        at org.apache.hadoop.ipc.Client.call(Client.java:1475)
        at org.apache.hadoop.ipc.Client.call(Client.java:1412)
        at org.apache.hadoop.ipc.ProtobufRpcEngine$Invoker.invoke(ProtobufRpcEngine.java:229)
        at com.sun.proxy.$Proxy15.create(Unknown Source)
        at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolTranslatorPB.create(ClientNamenodeProtocolTranslatorPB.java:296)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.apache.hadoop.io.retry.RetryInvocationHandler.invokeMethod(RetryInvocationHandler.java:191)
        at org.apache.hadoop.io.retry.RetryInvocationHandler.invoke(RetryInvocationHandler.java:102)
        at com.sun.proxy.$Proxy16.create(Unknown Source)
        at org.apache.hadoop.hdfs.DFSOutputStream.newStreamForCreate(DFSOutputStream.java:1648)
        at org.apache.hadoop.hdfs.DFSClient.create(DFSClient.java:1689)
        at org.apache.hadoop.hdfs.DFSClient.create(DFSClient.java:1624)
        at org.apache.hadoop.hdfs.DistributedFileSystem$7.doCall(DistributedFileSystem.java:448)
        at org.apache.hadoop.hdfs.DistributedFileSystem$7.doCall(DistributedFileSystem.java:444)
        at org.apache.hadoop.fs.FileSystemLinkResolver.resolve(FileSystemLinkResolver.java:81)
        at org.apache.hadoop.hdfs.DistributedFileSystem.create(DistributedFileSystem.java:459)
        at org.apache.hadoop.hdfs.DistributedFileSystem.create(DistributedFileSystem.java:387)
        at org.apache.hadoop.fs.FileSystem.create(FileSystem.java:911)
        at org.apache.hadoop.fs.FileSystem.create(FileSystem.java:892)
        at org.apache.hadoop.fs.FileSystem.create(FileSystem.java:789)
        at com.intenthq.pucket.Pucket$$anonfun$writeMeta$2$$anonfun$apply$10.apply(Pucket.scala:230)
        at com.intenthq.pucket.Pucket$$anonfun$writeMeta$2$$anonfun$apply$10.apply(Pucket.scala:230)
        at scalaz.$bslash$div$.fromTryCatchNonFatal(Either.scala:373)
        at com.intenthq.pucket.Pucket$$anonfun$writeMeta$2.apply(Pucket.scala:230)
        at com.intenthq.pucket.Pucket$$anonfun$writeMeta$2.apply(Pucket.scala:229)
        at scalaz.$bslash$div.flatMap(Either.scala:135)
        at com.intenthq.pucket.Pucket$.writeMeta(Pucket.scala:229)
        at com.intenthq.pucket.thrift.ThriftPucket$.create(ThriftPucket.scala:122)
        at com.intenthq.pucket.thrift.ThriftPucket$$anonfun$1.apply(ThriftPucket.scala:78)
        at com.intenthq.pucket.thrift.ThriftPucket$$anonfun$1.apply(ThriftPucket.scala:78)
```

**Comment by nathankleyn** [comment]
Fixed in #17! 👍
